### ISSUE TITLE
[zfb Migration Parity][Epic] Phase B-14 — Astro framework runtime + version-switcher + whitespace normalization in harness, plus zfb MDX named-import fix

### ISSUE THREAD

**Issue by Takazudo** [zudolab/zudo-doc#914]
- Issue: zudolab/zudo-doc#666 (Phase C parent)
- Super-epic: zudolab/zudo-doc#663

**Super-epic:** #663
**Super-epic base branch:** \`base/zfb-migration-parity\`
**This epic's base branch:** \`base/zfb-migration-parity-phase-b-14-harness-normalization\`

## Overview

Phase B-13 (#912, PR #913 merged 2026-04-30) cleared the four systematic causes from round 7:

- ✅ B-13-1 (strip-hidden-sidebar A-side asymmetry) — 24-route landmark cluster cleared
- ✅ B-13-2 (TOC island gating) — no-heading TOC over-render cleared
- ✅ B-13-3 (smart-quote / typography normalization) — 7-9 routes cleared
- ⚠️ B-13-4 (image-enlarge optimization) — files moved to public/ but path mismatch persists vs A's Astro-optimized `_astro/*.HASH.webp`

But the post-B-13 rerun shows **51 content-loss + 149 asset-loss = 200 regression-class routes** — same total as round 7. After applying simulated harness-level normalization for framework noise, **38 of 51** content-loss routes collapse, leaving only ~13 with real content differences. This proves the dominant remaining residual is *harness-level false positive*, not real content loss.

Filing as Phase B-14 sibling per #665's gating rule. Five sub-epics for the five new systematic causes.

### B-14-1 — Strip Astro framework runtime from `<main>` text extraction (harness)

**Problem**: Astro emits inline runtime scripts (`(()=>{var l=(n,t)=>{...self.Astro...astro:idle...})();`, `astro-island,astro-slot,astro-static-slot{display:contents}` style block, etc.) inside `<main>` whenever the page contains an `<astro-island>` element. The harness's text-extractor pulls these script bodies into the \`<main>\` text content (250–400 chars per route), driving the textHash diff. zfb has no equivalent inline runtime — its islands hydrate via `data-zfb-island` markers handled by a single shared script.

**Coverage**: All 51 content-loss routes have at least one of (\`(()=>{var l=\`, \`self.Astro\`, \`astro:idle\`, \`astro-island,astro-slot\`) inline in A's main extraction. Zero in B.

**Fix-locus**: \`scripts/migration-check/lib/normalize-html.mjs\` (or equivalent) — before computing textHash, strip:
- \`<script>\` element bodies inside \`<main>\` (Astro framework runtime)
- \`astro-island,astro-slot,astro-static-slot{display:contents}\` style declaration block
- Any \`(()=>{...self.Astro...})();\` IIFE pattern (defensive)

Conceptually identical to the #701 asset-loss carve-out — Astro framework noise should not be counted as content-loss.

**Acceptance**: Post-B-14-1 rerun shows ≥ 38 content-loss routes drop to identical (or move to asset-loss tier).

### B-14-2 — Strip version-switcher inline dropdown menu from both sides (harness)

**Problem**: A's Astro DocLayout renders \`<div class=\"version-switcher\">\` inline in \`<main>\` next to the breadcrumb. The element contains a button (\`Version: Latest\`) AND a hidden \`<ul>\` with version links (\`Latest\`, \`1.0.0\`, \`All versions\` — ~45 chars). zfb's host puts the version-switcher only in the header. The inline switcher is cosmetic chrome and the visible button text is functionally redundant with the header switcher.

**Coverage**: ~30 of 51 content-loss routes have \`Version: Latest\` in A's main extraction.

**Fix-locus**: \`scripts/migration-check/lib/strip-version-switcher.mjs\` (new) — strip \`<div class=\"version-switcher\" ...>...</div>\` (and elements with \`data-version-switcher\` attribute) from BOTH A and B before signal extraction. Same pattern as the existing strip-hidden-sidebar harness.

**Acceptance**: Post-B-14-2 rerun shows version-switcher text no longer contributes to textHash diffs.

### B-14-3 — Canonicalize whitespace (incl. `&nbsp;` / U+00A0) in text extraction (harness)

**Problem**: A's tag-listing pages render counts with \`&nbsp;\` (\`#ai &nbsp;(3)\`); B uses plain space (\`#ai (3)\`). After HTML entity decoding both should be equivalent, but the harness's text extraction preserves the byte difference (entity-encoded vs decoded; U+00A0 vs U+0020), flipping the textHash.

**Coverage**: 19 tag-listing routes (\`/docs/tags\`, \`/docs/tags/*\`, \`/ja/docs/tags\`, \`/ja/docs/tags/*\`).

**Fix-locus**: \`scripts/migration-check/lib/normalize-html.mjs\` — canonicalize all whitespace runs (incl. U+00A0 / \`&nbsp;\`) to a single ASCII space before computing textHash. Cheap, well-scoped, no false negatives.

**Acceptance**: Post-B-14-3 rerun shows 19 tag-listing routes drop to identical.

### B-14-4 — Canonicalize Astro `/_astro/<basename>.HASH.<ext>` asset paths (harness)

**Problem**: B-13-4 moved \`image-{small,wide,opt-out}.png\` to \`public/img/image-enlarge/*.webp\`. B's HTML now references these correctly. But A continues to reference Astro-optimized paths under \`/_astro/image-*.HASH.webp\` (Astro's image pipeline emits content-hashed filenames at build time). Same image content, different paths.

**Coverage**: 2 routes (\`/docs/components/image-enlarge\`, \`/ja/docs/components/image-enlarge\`).

**Fix-locus**: \`scripts/migration-check/lib/diff-artifacts.mjs\` (or asset-path canonicalization step) — recognize the \`/_astro/<filename>.HASH.<ext>\` Astro pattern and canonicalize the basename: \`image-wide.D1YdccyX_1EtJL4.webp\` → \`image-wide.webp\`. Compare against B's \`/img/image-enlarge/image-wide.webp\` after similarly stripping the parent path. May require a configurable allowlist of \"known A→B path remappings\" so the change doesn't accidentally collapse genuinely different asset paths.

**Acceptance**: Post-B-14-4 rerun shows the 2 image-enlarge routes drop to identical (or move to pure framework-noise carve-out).

### B-14-5 — zfb MDX named-import binding stripped from rendered output (zfb upstream)

**Problem**: When MDX source contains \`import { Island } from \"@takazudo/zfb\";\`, zfb's MDX → JSX emitter strips the \`{ Island }\` named binding but leaves the import skeleton. The malformed result \`import  from \"@takazudo/zfb\"\` is then rendered as a visible \`<p>\` paragraph in \`<main>\` (note the double-space between \`import\` and \`from\` from the stripped binding).

Verified on \`/docs/getting-started/setup-preset-generator\`:

```
<p>import PresetGenerator from '@/components/preset-generator';
import  from &quot;@takazudo/zfb&quot;;</p>
```

This is a zfb upstream parser bug. It's the same shape as the math-equations bridge fallback (B-12-3 was filed and fixed similarly).

**Coverage**: 2 routes (\`/docs/getting-started/setup-preset-generator\` EN + JA).

**Fix-locus**:
- Primary: \`@takazudo/zfb\` MDX → JSX emitter — handle \`import { name } from \"module\"\` the same way as \`import default from \"module\"\`. Both should be stripped from rendered output.
- After upstream fix: bump \`@takazudo/zfb\` version pin in this repo, re-run migration-check.
- Fallback if upstream fix is not feasible in this round: document as a known-divergence and exclude these 2 routes via harness allowlist; reopen once zfb ships a fix.

**Acceptance**: Post-B-14-5 rerun shows the 2 setup-preset-generator routes drop to identical (no more \`<p>import...</p>\` in rendered HTML).

## Acceptance criteria

After Phase B-14 ships:

- [ ] B-14-1 — strip Astro framework runtime; ≥ 38 content-loss routes drop
- [ ] B-14-2 — strip version-switcher menu DOM; remaining ~30 routes lose ~45-char delta
- [ ] B-14-3 — canonicalize whitespace; 19 tag-listing routes drop
- [ ] B-14-4 — canonicalize Astro image paths; 2 image-enlarge routes drop from non-carve-out asset-loss
- [ ] B-14-5 — zfb fixes named-import handling OR routes are allowlisted; 2 setup-preset-generator routes drop
- [ ] Post-B-14 rerun: residual content-loss + non-carve-out asset-loss ≤ 10 (Phase C inline-acceptance gate)

## Dependencies

- Phase B-13 (#912) merged ✅
- Phase C (#666) is currently paused awaiting B-14

## Notes

- Causes #1, #2, #3 are pure harness-side fixes (\`scripts/migration-check/lib/*\`) — no host-code changes needed.
- Cause #4 is also harness-side but may need a per-route allowlist to be safe.
- Cause #5 is the only cause requiring upstream zfb work; defer to allowlist if not feasible in this round.
- After B-14 ships, the genuine per-page residual that Phase C will handle should fit the \`5 < N ≤ 10\` band of #666's refinement rule (mostly source drift between origin/main and HEAD).

Full classification + per-cause coverage probe + reproduction recipe: \`docs/migration-check-reports/2026-04-30-phase-c-input-post-b13-analysis.md\` on branch \`base/zfb-migration-parity-phase-c-mop-up\`.

## Resumption (after B-14 merges)

    /x-wt-teams -s https://github.com/zudolab/zudo-doc/issues/666

**Comment by Takazudo** [comment]
🤖 Starting work on this epic in a Claude Code session (`/x-wt-teams` Super-Epic child). To avoid conflicts, please check the latest comments before starting another session on this epic.

**Comment by Takazudo** [comment]
## Step 1 complete — epic base + root PR created

- Epic base: `base/zfb-migration-parity-phase-b-14-harness-normalization`
- Root PR: https://github.com/zudolab/zudo-doc/pull/915 (draft)
- Targeting: `base/zfb-migration-parity` (super-epic base)

Next: create 4 worktrees + spawn child agents.

**Comment by Takazudo** [comment]
### topic-harness-strip-version-switcher — completed

Created `scripts/migration-check/lib/strip-version-switcher.mjs` (new file, mirrors `strip-hidden-sidebar.mjs` API: `hasVersionSwitcher`, `stripVersionSwitcher`, `maybeStripVersionSwitcher`). The module detects two markup shapes — A-side `<div class="version-switcher">` and B-side `<div data-version-switcher>` — and uses a balanced-depth div walker to remove the entire element including nested children. Added `stripVersionSwitcherDom = true` to `config.mjs`. Wired `maybeStripVersionSwitcher` into the `compareRoute` pipeline in `compare-routes.mjs` as the outer wrap after `maybeStripHiddenSidebar`. Includes 22 unit tests (detection, stripping, nested divs, multiple occurrences, symmetric stripping, enabled/disabled toggle); all pass. The gcoc review found no bugs or logic errors.

**Comment by Takazudo** [comment]
### topic-harness-canonicalize-asset-paths — completed

Added Astro asset path canonicalization to `scripts/migration-check/diff-artifacts.mjs`. The fix introduces `extractAstroCanonicalBasename()` and `buildAstroCanonicalMap()` that strip the content hash from `_astro/<name>.<HASH>.<ext>` filenames (e.g. `_astro/image-wide.D1YdccyX_1EtJL4.webp` → `image-wide.webp`). In `diffArtifacts()`, A-side `_astro/` paths are cross-matched against B-side paths by canonical basename when exactly one B-side candidate exists and no exact-path overlap occurs — so image-enlarge assets that moved to `img/image-enlarge/` in B are now treated as `present-in-both` and compared by file content rather than flagged as spurious `only-in-a` / `only-in-b` differences. Also added `.webp` and `.svg` to `BINARY_EXTENSIONS` (previously missing). 49 unit tests pass including new coverage for both helper functions.

**Comment by Takazudo** [comment]
### topic-zfb-named-import-fix — completed (Path A)

Took Path A (upstream zfb fix). Root cause: `markdown-rs`'s `MdxjsEsm` construct requires `mdx_esm_parse` to be `Some` to activate. `ParseOptions::mdx()` left it as `None`, so `import { Island } from "@takazudo/zfb"` fell through to Paragraph parsing. The `{ Island }` binding became an `MdxTextExpression` evaluating to `undefined` at runtime, which rendered as empty string — leaving the double-space skeleton `import  from "@takazudo/zfb"` visible in the page body.

Fix in **Takazudo/zudo-front-builder#94**: supply a permissive `mdx_esm_parse` (always returns `Ok`) so all import/export statements are classified as `MdxjsEsm` nodes. The emitter's `_ => String::new()` catch-all then silently drops them. Four regression tests added covering default import, named import, the two-consecutive-imports pattern (exact match for setup-preset-generator), and export statements.

zudo-doc2 PR **#916** documents the 2 affected routes in `pendingBinaryRebuildRoutes` in `scripts/migration-check/config.mjs`. Once the zfb PR merges and the binary is rebuilt (`cargo build -p zfb` + `pnpm install`), both routes should reach parity without further zudo-doc changes.

**Comment by Takazudo** [comment]
### topic-harness-normalize-html — completed

Implemented B-14-1 and B-14-3 as two new steps in scripts/migration-check/lib/normalize-html.mjs:

**B-14-1 (step 1b)**: Strips Astro framework runtime inline scripts and the astro-island/astro-slot display:contents style block from normalized HTML before signal extraction. Inline script elements (attribute-less, no src) whose body contains self.Astro, astro:idle, astro:load, or astro-island are removed entirely. The style block matching /astro-island,astro-slot/ is also stripped. This eliminates DOM-shape noise from domShapeHash (A has these elements, B does not) without affecting extractVisibleText which already strips script bodies.

**B-14-3 (step 2b)**: Canonicalizes &nbsp; HTML entities (case-insensitive) and U+00A0 non-breaking space characters to plain ASCII space. Tag-listing pages in Astro emit '#tag &nbsp;(N)' while zfb uses a plain space; the 6-char entity form inflates A's visible-text length and pushes B below the 95% content-loss threshold, causing false positives on all 19 tag-listing routes.

Both steps are covered by 12 new unit tests (7 for B-14-1, 5 for B-14-3). All 41 normalize-html tests pass. Pre-existing port-conflict flakiness in serve-snapshots.test.ts is unrelated.

**Comment by Takazudo** [comment]
## Step 8 — session report (workflow ends after merge step)

Implementation complete. All 5 sub-epics shipped. Root PR 915 marked ready. Awaiting mandatory epic-PR merge into super-epic base per Super-Epic child rule.

### Per-topic outcomes

| Sub-epic | Branch | Tests added | Status |
|----------|--------|-------------|--------|
| B-14-1 + B-14-3 (combined: same file) | …/harness-normalize-html → 11fecf0 | 14 | impl complete |
| B-14-2 | …/harness-strip-version-switcher → b51c660 | 22 | impl complete |
| B-14-4 | …/harness-canonicalize-asset-paths → ea17f9c | 4 | impl complete |
| B-14-5 | …/zfb-named-import-fix → 7f47696 (docs marker only on this side) | n/a | upstream PR https://github.com/Takazudo/zudo-front-builder/pull/94 + `pendingBinaryRebuildRoutes` documentation |

Post-review follow-up commit e2aa20d added by manager: extends B-14-3 to also canonicalize the `&#160;` (decimal) and `&#xa0;` (hex) numeric character references — gap caught by gcoc review.

### CI

- Build Doc History pass · Template Drift Check pass.
- Build Site / Type Check / E2E fail with `sh: 1: zfb: not found` — pre-existing infrastructure issue affecting the entire super-epic. Same pattern on the super-epic base itself (run 25164509105). Not introduced by Phase B-14.

### Test verification

- 112 unit tests pass across the 3 affected harness modules (43 normalize-html + 22 strip-version-switcher + 47 diff-artifacts).
- Per-route migration-check rerun deferred to Phase C (resumption command in next-step section).

### Acceptance criteria

All implementation ACs (B-14-1 through B-14-5) complete and unit-test-verified. The "Post-B-14 rerun: residual ≤ 10" final gate is owned by Phase C session.

Full session report: /home/takazudo/cclogs/zudo-doc2/20260430_222234-x-wt-teams-phase-b-14.md

Next: Super-Epic mandatory merge of PR 915 into `base/zfb-migration-parity`, then close issue and switch to super-epic base.

**Comment by Takazudo** [comment]
Workflow complete. Root PR: https://github.com/zudolab/zudo-doc/pull/915 (merged into base/zfb-migration-parity).
